File: shared/types/forecast.ts
// Mirrors backend/api/schemas/forecast.py

export interface DrawdownInfo {
  mean_max_drawdown: number;
  worst_case_drawdown: number;
}

export interface DistributionInfo {
  expected_return: number;
  median_return: number;
  ci_lower_90: number;
  ci_upper_90: number;
  prob_positive_return: number;
  worst_5pct: number;
  best_5pct: number;
}

export interface RiskAnalyticsInfo {
  value_at_risk_95: number;
  expected_shortfall_95: number;
  drawdown: DrawdownInfo;
  distribution: DistributionInfo;
}

export interface FanChartInfo {
  horizon_days: number[];
  percentiles: Record<string, number[]>; // e.g. "p5" -> values
}

export interface DensityInfo {
  bin_edges: number[];
  counts: number[];
}

export interface ModelParamsInfo {
  model_name: string;
  display_name: string;
  params: Record<string, number>;
}

export interface ExplanationInfo {
  summary: string;
  regime_description: string;
  regime_confidence_note: string;
  performance_basis: string;
  driving_features: string[];
}

export interface ForecastRequest {
  ticker: string;
  model_name?: string | null;
  n_sims?: number;
  horizon_days?: number;
  start?: string | null;
  end?: string | null;
}

export interface ForecastResponse {
  ticker: string;
  current_price: number;
  auto_selected: boolean;
  explanation: ExplanationInfo | null;
  model: ModelParamsInfo;
  risk_analytics: RiskAnalyticsInfo;
  fan_chart: FanChartInfo;
  density: DensityInfo;
}

File: backend/api/schemas/forecast.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field


class DrawdownInfo(BaseModel):
    mean_max_drawdown: float
    worst_case_drawdown: float


class DistributionInfo(BaseModel):
    expected_return: float
    median_return: float
    ci_lower_90: float
    ci_upper_90: float
    prob_positive_return: float
    worst_5pct: float
    best_5pct: float


class RiskAnalyticsInfo(BaseModel):
    value_at_risk_95: float
    expected_shortfall_95: float
    drawdown: DrawdownInfo
    distribution: DistributionInfo


class FanChartInfo(BaseModel):
    horizon_days: list[int]
    percentiles: dict[str, list[float]]
    sample_paths: list[list[float]] = Field(default_factory=list)


class DensityInfo(BaseModel):
    bin_edges: list[float]
    counts: list[float]


class ModelParamsInfo(BaseModel):
    model_name: str
    display_name: str
    params: dict[str, float]


class ExplanationInfo(BaseModel):
    summary: str
    regime_description: str
    regime_confidence_note: str
    performance_basis: str
    driving_features: list[str]


class ForecastRequest(BaseModel):
    ticker: str
    model_name: str | None = None  # None = auto-select via ModelSelectionEngine
    n_sims: int = 5000
    horizon_days: int = Field(default=63, ge=1, le=252)
    start: date | None = None
    end: date | None = None


class ForecastResponse(BaseModel):
    ticker: str
    current_price: float
    auto_selected: bool
    explanation: ExplanationInfo | None = None  # populated only when auto-selected
    model: ModelParamsInfo
    risk_analytics: RiskAnalyticsInfo
    fan_chart: FanChartInfo
    density: DensityInfo
